tests/test_core.py

```diff
@@ -96,6 +86,166 @@ def test_block_iterators(self):
+        self.invalid_net("Outputs cannot be arguments", net)
+
+        wrong_ops = ('%', '!', 'a', 'f', '<<', '>>', '&&', '||', '==')
+        for op in wrong_ops:
+            net = self.new_net(op=op)
+            self.invalid_net("not from acceptable set", net)
+
+    def test_net_wrong_num_args(self):
+        for op in 'w~rsm':
+            net = self.new_net(op=op)
+            self.invalid_net("op only allowed 1 argument", net)
+        for op in '&|^n+-*<>=':
+            net = self.new_net(op=op, args=3)
+            self.invalid_net("op only allowed 2 arguments", net)
+        for op in 'x@':
+            net = self.new_net(op=op, args=4)
+            self.invalid_net("op only allowed 3 arguments", net)
+
+    def test_net_wrong_bitwidth(self):
+        net = self.new_net(op='x', args=tuple(pyrtl.Input(i) for i in range(1, 4)))
+        self.invalid_net("args have mismatched bitwidths", net)
+        net = self.new_net(op='x', args=tuple(pyrtl.Input(2) for i in range(3)))
+        self.invalid_net("mux select must be a single bit", net)
+
+        for op in '&|^n+-*<>=':
+            net = self.new_net(op=op, args=(tuple(pyrtl.Input(i) for i in range(2, 4))))
+            self.invalid_net("args have mismatched bitwidths", net)
+
+        net = self.new_net(op='m', op_param=(1234, pyrtl.MemBlock(2, 3)), args=1)
+        self.invalid_net("mem addrwidth mismatch", net)
+
+        net = self.new_net(op='@', op_param=(1234, pyrtl.MemBlock(2, 2)),
+                           args=tuple(pyrtl.Input(i) for i in (4, 2, 1)))
+        self.invalid_net("mem addrwidth mismatch", net)
+        net = self.new_net(op='@', op_param=(1234, pyrtl.MemBlock(2, 2)),
+                           args=tuple(pyrtl.Input(i) for i in (2, 4, 1)))
+        self.invalid_net("mem bitwidth mismatch", net)
+        net = self.new_net(op='@', op_param=(1234, pyrtl.MemBlock(2, 2)), args=3)
+        self.invalid_net("mem write enable must be 1 bit", net)
+
+    def test_net_wrong_num_op_params(self):
+        for op in 'w~r':
+            net = self.new_net(op=op, op_param=('hi', 'how', 'r', 'u'), args=1)
+            self.invalid_net("op_param should be None", net)
+        for op in '&|^n+-*<>=c':
+            net = self.new_net(op=op, op_param=('hi', 'how', 'r', 'u'), args=2)
+            self.invalid_net("op_param should be None", net)
+        net = self.new_net(op='x', op_param=('hi', 'how', 'r', 'u'),
+                           args=tuple(pyrtl.Input(i) for i in (1, 2, 2)))
+        self.invalid_net("op_param should be None", net)
+
+    def test_net_wrong_op_param_select(self):
+        net = self.new_net(op='s', op_param='hi', args=1)
+        self.invalid_net("select op requires tuple op_param", net)
+        net = self.new_net(op='s', op_param=(-2,), args=1)
+        self.invalid_net("op_param out of bounds", net)
+        net = self.new_net(op='s', op_param=(10,), args=1)
+        self.invalid_net("op_param out of bounds", net)
```

Comment:
could we have an test that tests tuples with non-int values such as `True`, and also a test to make sure that subsequent values in a tuple throw errors even if the initial one is valid?